Improve docs for abi required-ness (#3137)
fubuloubu/web3.py: docs/web3.eth.rst

.. py:method:: Eth.contract(address=None, contract_name=None, ContractFactoryClass=Contract, **contract_factory_kwargs)

>     An ``abi`` specifies how a contract can be interacted with. The ``abi`` 

Reviewer: I feel like this clarification should go down below, as in the "following arguments are accepted" section. This first section discusses the actual named arguments associated with the function, in order. `abi`, while necessary for making contract calls, isn't a named arg, just an optional kwarg (whether this is a good design is perhaps another question :thinking: ) Maybe adding it as a description to the optional kwargs list is a better place? That then opens it as a place for describing usage of the others in that list. 

Just my feels, open for discussion!

Author: Would be ideal to call something out here in the contract docs. I wonder if it would help to include abi as an argument in the method signature above, or some way of indicating it is necessary for the interface to work.

Agree this paragraph doesnt seem quite right in the contract API docs. I think it might clutter the list down below, but there is a link to the Contracts guide from there. I think the paragraph might be a good replacement for the contract docs under `Contract.abi`. What do you think?

Reviewer: Agree with putting more info under `Contract.abi`. Could even explain what ABI stands for there. I also agree with some indication here that abi is required if you want to interact with the contract, maybe just not first thing?

    The following arguments are accepted for contract class creation.

>     :param abi: Application Binary Interface. Usually provided since an ``abi`` is required to interact with any contract.

Author: @fselmo @pacrob I like the sphinx `param` block! I'm not sure about all of these params, but I took my best guess at as many as I could. Are there any words we can add or change here?

Thanks!

    :type abi: ABI
    :param asm: Asssembly code generated by the compiler
    :param ast: Abstract Syntax Tree of the contract generated by the compiler
    :param bytecode: Bytecode of the contract generated by the compiler
    :param bytecode_runtime: Bytecode stored at the contract address, excludes the constructor and initialization code
    :param clone_bin:
    :param dev_doc:
    :param decode_tuples: Optionally convert tuples/structs to named tuples
    :param interface:
    :param metadata: Contract Metadata generated by the compiler

>     :param opcodes: Opcodes of the contracts generated by the compiler

Reviewer 2: ```suggestion
    :param opcodes: Opcodes for the contract generated by the compiler
```

Is this meant to be singular here?

    :param src_map:
    :param src_map_runtime:
    :param user_doc:
    :return: Instance of the contract
    :rtype: Contract

>     :raises TypeError: if the address is not provided

Reviewer 2: ```suggestion
    :raises TypeError: If the address is not provided
```
 Everything else seems to be capitalized here.

>     :raises AttributeError: if the contract class is not initialized

Reviewer 2: ```suggestion
    :raises AttributeError: If the contract class is not initialized
```
